### PR TITLE
Mark v1beta1 apis as deprecated

### PR DESCRIPTION
**Pull request by lbernick**
Now that v1 versions of Task, TaskRun, Pipeline, and PipelineRun are available and used as the storage version of these APIs, this commit marks the v1beta1 versions of these APIs as deprecated.

This commit also clarifies that "support" for these APIs during the deprecation period means that they will receive bug fixes but not new features.

/kind documentation
Part of https://github.com/tektoncd/pipeline/issues/5541

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- n/a Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
action required: v1beta1 Task, TaskRun, Pipeline, and PipelineRun APIs are deprecated and will be removed. Please migrate to the v1 versions of these APIs following the migration guide at https://github.com/tektoncd/pipeline/blob/main/docs/migrating-v1beta1-to-v1.md.
```

### CODE REVIEW

**Review by Yongxuanzhang — COMMENTED** [review]
/lgtm

**Review by QuanZhang-William — COMMENTED** [review]
Thanks, Lee! Changes look good to me. I'm curious about what is the impact of marking the `v1beta1` crds as `deprecated`? Does it mean I can no longer `kubectl apply` a `v1beta1 pipelinerun` for example?

**Review by jerop — APPROVED** [review]
/test pull-tekton-pipeline-go-coverage

**Comment by afrittoli** [comment]
> Thanks, Lee! Changes look good to me. I'm curious about what is the impact of marking the `v1beta1` crds as `deprecated`? Does it mean I can no longer `kubectl apply` a `v1beta1 pipelinerun` for example?

There is no change until the CRD is "removed" (i.e. not served anymore).
It signals users that they should migrate their resources to `v1`.

**Review by afrittoli — COMMENTED** [review]
Thank you!
/approve

File: api_compatibility_policy.md
Comment (pritidesai) [inline api_compatibility_policy.md:48]: Can we please add a timeline of when this deprecation was announced and until when these API will receive bug fixes? It will help the users to understand and plan their upgrade. Thanks!

File: docs/deprecations.md
Comment (pritidesai) [inline docs/deprecations.md:28]: Is this accurate? The deprecation is announced in 0.50 which is LTS and will have the estimated end of the life in a year around `2024-07-20` (might be v0.62.0) unless we cut release v1 before that.

Comment (lbernick) [reply docs/deprecations.md:28]: Currently our policy for beta CRDs is 9 months notice, or 9 releases. So the last release with support would be 0.59.0, which would have an end of life of 1 yr + 9 months from the date 0.50.0 is released. I think having the last release with support listed here makes sense-- if we put a date I'm worried people might think we cannot remove the CRDs until that date.

File: api_compatibility_policy.md
Comment (lbernick) [reply api_compatibility_policy.md:48]: Is there info missing from docs/deprecations.md that would be helpful here?

Comment (pritidesai) [reply api_compatibility_policy.md:48]: There is no missing info from `docs/deprecations.md` but at the same time there is no harm in adding a bit more details here for the users. As a user, how would I know when these changes were introduced if I am just looking at the [api_compatibility_policy.md#crd-api-versions](https://github.com/tektoncd/pipeline/blob/86ff3643242481624eed6dd35ad95b63d1a6a01e/api_compatibility_policy.md#crd-api-versions). There is no mention of the `docs/deprecations.md`. At the least if we add a link to the deprecations.md here, it will be very helpful for the users.

File: docs/deprecations.md
Comment (pritidesai) [reply docs/deprecations.md:28]: I am not sure if I am following, 1 year + 9 months from the date 0.50 is released. 

Yup our policy for beta CRDs is 9 months/releases.

Let's take walk through this scenario. There is a bug discovered in 0.50 with `v1beta1` by the time we are about to release 0.60 (May, 2024). Do we keep `v1beta1` CRDs around in `0.60` to fix such potential bugs or those CRDs are deleted in `0.59`? The way we are cherry picking after fixing an issue in `main`, are we going to follow the same process? If we are going to follow the same process and if we delete `v1beta1` CRDs, how can we fix a bug in `0.50`? (lot of ifs 🤣) Does it make sense? Am I complicating it?

File: api_compatibility_policy.md
Comment (lbernick) [reply api_compatibility_policy.md:48]: added a link! thanks for the feedback.

File: docs/deprecations.md
Comment (lbernick) [reply docs/deprecations.md:28]: > There is a bug discovered in 0.50 with v1beta1 by the time we are about to release 0.60 (May, 2024). Do we keep v1beta1 CRDs around in 0.60 to fix such potential bugs or those CRDs are deleted in 0.59?

If we want to backport a fix to 0.50, we could just create a commit on the release-v0.50.x branch. I don't think we need to keep the CRDs around on the main branch in case we need to backport a bug fix.

I'm curious if there's confusion here due to the client libraries + go structs, vs the actual CRD and whether it's served? In https://github.com/tektoncd/pipeline/pull/6768, I updated our policy to say that we'd keep the Go clients for v1beta1 CRDs until the last release when they are served (v0.59.0) reaches its EOL (1 year later, or when we are releasing v0.69.0).

**Comment by pritidesai** [comment]
@lbernick and I talked offline, thanks @lbernick for this PR 🙏 

Once the PR is updated to include the steps you shared, we can merge this.

```
Release 0.50 where v1beta1 is deprecated
Release v0.59 where v1beta1 is no longer served, but we still have the client libraries
Release 0.71 we remove the client libraries
```

/approve

**Comment by lbernick** [comment]
> @lbernick and I talked offline, thanks @lbernick for this PR 🙏
> 
> Once the PR is updated to include the steps you shared, we can merge this.
> 
> ```
> Release 0.50 where v1beta1 is deprecated
> Release v0.59 where v1beta1 is no longer served, but we still have the client libraries
> Release 0.71 we remove the client libraries
> ```

Done!

**Comment by vdemeester** [comment]
So this mean we are planning to remove `v1beta1` in less than a year (9 month from now), am I right ? This will mean we are removing `v1beta1` API "quicker" than we removed `v1alpha1` API I feel, which is a bit worrysome 😅 (especially seeing that few of our customers *just* moved from `v1alpha1` to `v1beta1`). We gave more than a year to migrate from `v1alpha1` to `v1beta1`, but we give *less* than a year to migrate from `v1beta1` to `v1` ? 

All our "plumbing" (here in `tekton` or in `tektoncd/plumbing`) are still using `v1beta1` even. I would *assume* we would at least have them all migrated to `v1` (and have a few cycle of "uses" with them in `v1`) before marking `v1beta1` as deprecated. A discussion/debate could also take place of the *catalogs* on the same subject (`v1beta1` API still served there).

So.. I am really worried about marking it as deprecated it as early as 0.50 (July release).

/hold

**Comment by lbernick** [comment]
> So this mean we are planning to remove `v1beta1` in less than a year (9 month from now), am I right ? This will mean we are removing `v1beta1` API "quicker" than we removed `v1alpha1` API, which is a bit worrysome 😅 (especially seeing that few of our customers _just_ moved from `v1alpha1` to `v1beta1`). We gave more than a year to migrate from `v1alpha1` to `v1beta1`, but we give _less_ than a year to migrate from `v1beta1` to `v1` ?

If we want to allow more time to migrate from v1beta1 to v1, we should discuss changing our compatibility policy. If we don't follow our compatibility policy, how will we know when it's acceptable to remove support for these CRDs?

I also just want to note a few things:
- our deprecation table describes the "earliest" date or release of removal, but we still have the option to remove the v1beta1 CRDs later
- v1 CRDs have already been available since v0.43.0
- If you want longer support for v1beta1 CRDs, you can stick with v0.59.0 (which is LTS) for a year rather than upgrading

> In addition, all our "plumbing" (here in `tekton` or in `tektoncd/plumbing`) are still using `v1beta1` even. I would _assume_ we would at least have them all migrated to `v1` (and have a few cycle of "uses" with them in `v1`) before marking `v1beta1` as deprecated (and we also need to actually exercise a recent version on tektoncd/pipeline ourselves.. our cluster still runs 0.44). A discussion/debate could also take place of the _catalogs_ on the same subject (`v1beta1` API still served there).

That's a good point; it could definitely make sense to have some usage of v1 in plumbing before we mark v1beta1 as deprecated. We can also upgrade the 4 tasks in tektoncd-catalog but I don't think we should wait for the tasks in tektoncd/catalog to upgrade. (However, I'd hope our integration tests and upgrade tests are good enough to give us sufficient confidence in the v1 api!)

**Comment by lbernick** [comment]
@vdemeester I've opened https://github.com/tektoncd/pipeline/pull/6923 for discussion on the beta CRD support window.

**Comment by afrittoli** [comment]
@vdemeester @lbernick shall we try and get this merged for v0.50?

**Comment by vdemeester** [comment]
@afrittoli I would say yes 🙃 
@lbernick I approved https://github.com/tektoncd/pipeline/pull/6923 👼🏼

**Comment by lbernick** [comment]
/hold cancel
I've updated this PR based on the discussions resolved in https://github.com/tektoncd/pipeline/pull/6923

**Comment by vdemeester** [comment]
@lbernick needs a rebase 🙏🏼

**Comment by lbernick** [comment]
done! just needs lgtm

**Pull request merged**
